<a href="https://colab.research.google.com/github/Mehrabwv/python/blob/main/Welcome_to_Colaboratory.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
weight1= input('What is your weight now?: ')
increase1= input('How much increase do you expect?: ')
year1= input('In how many years?: ')

weight = int(weight1)
increase = int(increase1)
yeare = int(year1)
year2 = yeare


for year2 in range(0,year2):
  moonweight=(weight+(increase*year2))*0.165
  print(year2,moonweight)

What is your weight now?: 1
How much increase do you expect?: 1
In how many years?: 1
0 0.165


# ***Python For Kids***   


# Chapter 1 - Not all snakes slither
# (Basics)

In [ ]:
print('How are you?')

How are you?


In [ ]:
print("hello world")

hello world


In [ ]:
print('''hello world''')

hello world


#Chapter 2 - Calculations and Variables
#Calculations

In [ ]:
8*3.57   #8 times 3.57 = 28.56

28.56

Pg 16 - puzzle

In [ ]:
10*365

3650

In [ ]:
20+3650

3670

In [ ]:
3*52

156

In [ ]:
3670-156

3514

BODMAS


add
5 to 30, and then multiply the result
by 20

In [ ]:
(5+30)*20

700

 Adding 5 to 30 gives 35.

 Multiplying 35 by 20 gives 700.

 Dividing 700 by 10 gives the final answer of 70.

In [ ]:
 ((5 + 30) * 20) / 10

70.0

Variables

In [ ]:
fred=100
print(fred)

100


In [ ]:
fred=100
john=fred+5
print(john)

105


In [ ]:
# You cant start a variable with a number!
number_of_coins = 200
print(number_of_coins)

200


Pg 21 - more info

In [ ]:
 20 + 10 * 365 - 3 * 52

3514

In [ ]:
found_coins = 20
magic_coins = 10
stolen_coins = 3

found_coins + magic_coins * 365 - stolen_coins * 52

3514

In [ ]:
found_coins = 20
magic_coins = 10

stolen_coins = 2   # Very easy to change the variable if neeeded

found_coins + magic_coins * 365 - stolen_coins * 52

3566

# Chapter 3 - Strings, Lists, Tuples, and Maps

A string is a text

In [ ]:
fred = "What is 1 + 1?"
print(fred)

What is 1 + 1?


In [ ]:
#To create multiple line string, use '''
fred = '''How do dinosaurs pay their bills?
With tyrannosaurus checks!'''
print(fred)

How do dinosaurs pay their bills?
With tyrannosaurus checks!


In [ ]:
silly_string = '''He said, "Aren't can't shouldn't wouldn't."'''
print(silly_string)

He said, "Aren't can't shouldn't wouldn't."


In [ ]:
single_quote_str = 'He said, "Aren\'t can\'t shouldn\'t wouldn\'t."'
print(single_quote_str)

He said, "Aren't can't shouldn't wouldn't."


In [ ]:
myname = 'mehrab'
string = 'hello my name is %s '    #Dont forget to put %s or error will occur
print(string % myname)

hello my name is mehrab 


In [ ]:
nock = 'what is %s + %s?'
print(nock %(1,5))

what is 1 + 5?


Multiplying strings  - pg 31

In [ ]:
print(10*'hello ')

hello hello hello hello hello hello hello hello hello hello 


Strings

In [ ]:
mylist = 'water, food, air, games, house, bed, tv, soap'     # This is a variable
print(mylist)

water, food, air, games, house, bed, tv, soap


In [ ]:
mylist = ['water', 'food', 'air', 'games', 'house', 'bed', 'tv', 'soap']
print(mylist)

['water', 'food', 'air', 'games', 'house', 'bed', 'tv', 'soap']


In [ ]:
mylist = ['water', 'food', 'air', 'games', 'house', 'bed', 'tv', 'soap']
print(mylist[2])    # Although it says no.2 , the no. 3 item is shown at the list starts with 0

air


In [ ]:
mylist = ['water', 'food', 'air', 'games', 'house', 'bed', 'tv', 'soap']
print(mylist[0]) 

water


In [ ]:
mylist = ['water', 'food', 'air', 'games', 'house', 'bed', 'tv', 'soap']

mylist[2] = 'oxygen'     # It is very easy to change the item (from air to oxygen)     -> Replacing
print(mylist)

['water', 'food', 'oxygen', 'games', 'house', 'bed', 'tv', 'soap']


In [ ]:
mylist = ['water', 'food', 'air', 'games', 'house', 'bed', 'tv', 'soap']
print(mylist[1:4])   

['food', 'air', 'games']


In [ ]:
nums = ['1','3','5','7','9']
words =['one','three','five','seven','nine']
print(nums + words)

['1', '3', '5', '7', '9', 'one', 'three', 'five', 'seven', 'nine']


In [ ]:
nums = ['1','3','5','7','9']
words =['one','three','five','seven','nine']
print(nums[1] + words[1])

3three


In [ ]:
mylist = ['water', 'food', 'air', 'games', 'house', 'bed', 'tv', 'soap']

mylist.append('cook')      # It is very easy to add items to the list
print(mylist)

['water', 'food', 'air', 'games', 'house', 'bed', 'tv', 'soap', 'cook']


In [ ]:
mylist = ['water', 'food', 'air', 'games', 'house', 'bed', 'tv', 'soap']

del mylist[7]     # It is also very easy to delete a item from the list
print(mylist)

['water', 'food', 'air', 'games', 'house', 'bed', 'tv']


In [ ]:
mylist = ['water', 'food', 'air', 'games', 'house', 'bed', 'tv', 'soap']
nums = [1,2,3,4]
add = nums + mylist

print(add)

[1, 2, 3, 4, 'water', 'food', 'air', 'games', 'house', 'bed', 'tv', 'soap']


In [ ]:
words = ['a','b']

print(words * 5)

['a', 'b', 'a', 'b', 'a', 'b', 'a', 'b', 'a', 'b']


note that you cannot divide or subtract a list 

and you can only add a list to a list but not another variable, number or word

**Tuples are like lists, however items cannot be added or removed with a code**

In [ ]:
what = (0,1,2,3,4)    # It is used because it is sometimes more useful to not be 
print(what[3])        # able to change items inside a tuples easily

3


Python map

In [ ]:
favorite_sports = ['Ralph Williams, Football',    # This is a list
 'Michael Tippett, Basketball',
 'Edward Elgar, Baseball',
 'Rebecca Clarke, Netball',
 'Ethel Smyth, Badminton',
 'Frank Bridge, Rugby']

print(favorite_sports)

['Ralph Williams, Football', 'Michael Tippett, Basketball', 'Edward Elgar, Baseball', 'Rebecca Clarke, Netball', 'Ethel Smyth, Badminton', 'Frank Bridge, Rugby']


In [ ]:
 favorite_sports = {'Ralph Williams' : 'Football',   # This is a map
 'Michael Tippett' : 'Basketball',
 'Edward Elgar' : 'Baseball',
 'Rebecca Clarke' : 'Netball',
 'Ethel Smyth' : 'Badminton',    # Uses {}
 'Frank Bridge' : 'Rugby'}

 print(favorite_sports)

{'Ralph Williams': 'Football', 'Michael Tippett': 'Basketball', 'Edward Elgar': 'Baseball', 'Rebecca Clarke': 'Netball', 'Ethel Smyth': 'Badminton', 'Frank Bridge': 'Rugby'}


In [ ]:
 favorite_sports = {'Ralph Williams' : 'Football',  
 'Michael Tippett' : 'Basketball',
 'Edward Elgar' : 'Baseball',
 'Rebecca Clarke' : 'Netball',
 'Ethel Smyth' : 'Badminton',    # You can easily find the sport
 'Frank Bridge' : 'Rugby'}

 print(favorite_sports['Rebecca Clarke'])

Netball


In [ ]:
favorite_sports = {'Ralph Williams' : 'Football',  
 'Michael Tippett' : 'Basketball',
 'Edward Elgar' : 'Baseball',
 'Rebecca Clarke' : 'Netball',
 'Ethel Smyth' : 'Badminton',    
 'Frank Bridge' : 'Rugby'}

del favorite_sports['Frank Bridge']    # Easy to delete a name/item

print(favorite_sports)

{'Ralph Williams': 'Football', 'Michael Tippett': 'Basketball', 'Edward Elgar': 'Baseball', 'Rebecca Clarke': 'Netball', 'Ethel Smyth': 'Badminton'}


In [ ]:
favorite_sports = {'Ralph Williams' : 'Football',  
 'Michael Tippett' : 'Basketball',
 'Edward Elgar' : 'Baseball',
 'Rebecca Clarke' : 'Netball',
 'Ethel Smyth' : 'Badminton',    
 'Frank Bridge' : 'Rugby'}

favorite_sports['Ethel Smyth']= 'Hockey'

print(favorite_sports)

{'Ralph Williams': 'Football', 'Michael Tippett': 'Basketball', 'Edward Elgar': 'Baseball', 'Rebecca Clarke': 'Netball', 'Ethel Smyth': 'Hockey', 'Frank Bridge': 'Rugby'}


In [ ]:
favorite_sports = {'Ralph Williams' : 'Football',  
 'Michael Tippett' : 'Basketball',
 'Edward Elgar' : 'Baseball',
 'Rebecca Clarke' : 'Netball',
 'Ethel Smyth' : 'Badminton',    
 'Frank Bridge' : 'Rugby'}           # Adding two maps does not work

favorite_colors = {'Malcolm Warner' : 'Pink polka dots',
 'James Baxter' : 'Orange stripes',
 'Sue Lee' : 'Purple paisley'}

# print(favorite_sports + favorite_colors)

Programming puzzles (pg 41)



***1: Favorites***

Make a list of your favorite hobbies and give the list the variable
name games. Now make a list of your favorite foods and name the
variable foods. Join the two lists and name the result favorites.
Finally, print the variable favorites.



In [ ]:
games = ['videogames','badminton','cycling']
foods = ['pizza','burger','friedchicken']
favorites = games + foods
print(favorites)

['videogames', 'badminton', 'cycling', 'pizza', 'burger', 'friedchicken']


***2: Counting Combatants***

If there are 3 buildings with 25 ninjas hiding on each roof and
2 tunnels with 40 samurai hiding inside each tunnel, how many
ninjas and samurai are about to do battle? (You can do this with
one equation in the Python shell.)

In [ ]:
print((3*25)+(2*40))

155


***3: Greetings!***

Create two variables: one that points to your first name and
one that points to your last name. Now create a string and use
placeholders to print your name with a message using those two
variables, such as “Hi there, Brando Ickett!”

In [ ]:
name1 ='Mehrab'
name2 ='Rahman'
string ='Hi there, %s %s'
print(string %(name1,name2) )

Hi there, Mehrab Rahman


#Chapter 4 - Drawing  with turtles

 ***Skipped***    - hard to use turtle in colab

# Chapter 5 - Asking Questions with if and else

If Statments

In [ ]:
age = 14       # This is an 'if' statement
if age > 10:      # Don't forgetg to add the semicolon :
  print('you are very old indeed')     # Always leave four spaces/tab before print or else
                                       # syntax error will occur as the code does not alline

you are very old indeed


In [ ]:
 age = 9

 if age < 10:
   print('You are too young!')
   print('You shall not pass!')
   print('Why are you still here?')

You are too young!
You shall not pass!
Why are you still here?


In [ ]:
age = 10
if age == 10:
  print('perfect!')

perfect!


In [ ]:
age = 14
if age >= 14:
  print('What is 123+123')

What is 123+123


***If-then-Else Statement***

In [ ]:
age = 10
if age <= 10:
  print('You are too young')
else:                                     # Dont forget : after 'else'
  print('You can keep on readinng this')

You are too young


In [ ]:
age = 14
if age <= 10:
  print('You are too young')
else:                                     
  print('You can keep on readinng this')

You can keep on readinng this


***if and elif Statements***

elif means else if.      
This is used as elif can be repeated unlike if

In [ ]:
joke = 0                   # Joke list

if joke == 1:
  print('What’s the best thing about Switzerland?')
  print('I don’t know, but the flag is a big plus.')
elif joke == 2:
  print('Why do we tell actors to “break a leg?”')
  print('Because every play has a cast.')
elif joke == 3:
  print('Heard about the new restaurant called Karma?')
  print('There’s no menu: You get what you deserve.')
elif joke == 4:
  print('Did you hear about the actor who fell through the floorboards?')
  print('He was just going through a stage.')
elif joke == 5:
  print('Did you hear about the claustrophobic astronaut?')
  print('He just needed a little space.')
elif joke == 6:
  print('''Why don’t scientists trust atoms? 
Because they make up everything.''')
elif joke == 7:
  print('''Why should the number 288 never be mentioned?
It’s two gross. - 1 gross is 144''')
else: 
  print('Joke list from 1 to 7')

Joke list from 1 to 7


You can use 'or' and 'and' to produce simpler and shorter codes

In [ ]:
age = 10       

if age == 10 or age == 11 or age == 12 or age == 13:    
  print('What is 13 + 49 + 84 + 155 + 97? A headache!')
else:
  print('You do not meet the age requirements!')

What is 13 + 49 + 84 + 155 + 97? A headache!


In [ ]:
age =  12

if age >= 10 and age <= 15:
  print('So you are interested to learn coding?')
else:
  print('Sorry your age does not meet the requirements :(')

So you are interested to learn coding?


0 is a number. None is not a number

In [ ]:
num = None
print(num)

None


In [ ]:
num = None
if num == None:
  print('None does not have a value')

None does not have a value


You cannot add a string to if

In [ ]:
num = 10
if num == 10: 
  print('This is a number')

This is a number


In [ ]:
num = '10'      # Nothing runs as this is a string and not a number 
if num == 10:        
  print('This is a string')

To solve this problem, we need to convert it using *int*.  This is useful in the future.   *int* converts a string into a number

In [ ]:
num = '10'      
new_num = int(num)    # Use int(-)

if new_num == 10:        
  print('This is a string')

This is a string


To convert a number into a string, we use *str*

In [ ]:
num = 10
new_num = str(num)

print(new_num + ' wow')   # Numbers and strings dont go together so we can see that new num is a string

10 wow


However, you will get an error if you try to convert a decimal as it only accepts integer(whole no.). So you must use float

In [ ]:
num = '10.5'
new_num = float(num)
print(new_num)

10.5


You will get a ValueError if you try converting a string that does not contain a number

***Programming puzzle - pg 65***

***1: Are You Rich?***

What do you think the following code will do? Try to figure out
the answer without typing it into the shell, and then check your
answer.

In [ ]:
money = 2000
if money > 1000:
 print("I'm rich!!")
else:
 print("I'm not rich!!")
 print("But I might be later...")

I'm rich!!


The programme checks if you are rich. If you have more than 1000 money, you are considered rich.

***2: Twinkies!***

Create an if statement that checks whether a number of Twinkies
(in the variable twinkies) is less than 100 or greater than 500. Your
program should print the message “Too few or too many” if the
condition is true.

In [ ]:
twinkies = 500
if twinkies < 100 or twinkies > 500:
  print('Too few or too many')

***3: Just the Right Number***

Create an if statement that checks whether the amount of money
contained in the variable money is between 100 and 500 or between
1,000 and 5,000.

In [ ]:
money = 1005
if money > 100 and money < 500:
  print('Money between 100 and 500')
elif money > 1000 and money < 5000:
  print('Money between 1000 and 5000')

Money between 1000 and 5000


***4: I Can Fight Those Ninjas***
Create an if statement that prints the string “That’s too many”
if the variable ninjas contains a number that’s less than 50, prints
“It’ll be a struggle, but I can take ’em” if it’s less than 30, and
prints “I can fight those ninjas!” if it’s less than 10. You might
try out your code with: ninjas = 5

In [ ]:
ninjas = 9
if ninjas < 50 and ninjas >= 30:
  print('That’s too many')
elif ninjas < 30 and ninjas >= 10:
  print('''It’ll be a struggle, but I can take ’em''')
elif ninjas < 10:
  print('I can fight those ninjas!')

I can fight those ninjas!


# Chapter 6 - Going loopy

for loops repeat a set of codes a number of times until it is complete. This can help us reduce the amount of codes we need to write if we need to repeat a set of codes multiple times.

In [ ]:
for x in range(0,5):   # Don't forget :
  print('hello')

hello
hello
hello
hello
hello


In [ ]:
print(list(range(10,20)))

[10, 11, 12, 13, 14, 15, 16, 17, 18, 19]


The code is telling Python to 

start counting from 0 and stop before reaching 5 and 

for each number we count, store the value in the variable x.

In [ ]:
for x in range(0,5):
  print('hello %s' % x)

hello 0
hello 1
hello 2
hello 3
hello 4


In [ ]:
mylist = ['water', 'food', 'air', 'games', 'house', 'bed', 'tv', 'soap']
for i in mylist:    # The code is asking Python to store each mylist variable in i
  print(i)

water
food
air
games
house
bed
tv
soap


In [ ]:
foodlist = ['water','mango','apple']
for i in foodlist:
  print(i)    # This command prints two of each item in the list
  print(i)    # Dont forget about indentation error

water
water
mango
mango
apple
apple


In [ ]:
found_coins = 20
magic_coins = 70
stolen_coins = 3
coins = found_coins
for week in range(1, 53):
  coins = coins + magic_coins - stolen_coins
  print('Week %s = %s' % (week, coins))

Week 1 = 87
Week 2 = 154
Week 3 = 221
Week 4 = 288
Week 5 = 355
Week 6 = 422
Week 7 = 489
Week 8 = 556
Week 9 = 623
Week 10 = 690
Week 11 = 757
Week 12 = 824
Week 13 = 891
Week 14 = 958
Week 15 = 1025
Week 16 = 1092
Week 17 = 1159
Week 18 = 1226
Week 19 = 1293
Week 20 = 1360
Week 21 = 1427
Week 22 = 1494
Week 23 = 1561
Week 24 = 1628
Week 25 = 1695
Week 26 = 1762
Week 27 = 1829
Week 28 = 1896
Week 29 = 1963
Week 30 = 2030
Week 31 = 2097
Week 32 = 2164
Week 33 = 2231
Week 34 = 2298
Week 35 = 2365
Week 36 = 2432
Week 37 = 2499
Week 38 = 2566
Week 39 = 2633
Week 40 = 2700
Week 41 = 2767
Week 42 = 2834
Week 43 = 2901
Week 44 = 2968
Week 45 = 3035
Week 46 = 3102
Week 47 = 3169
Week 48 = 3236
Week 49 = 3303
Week 50 = 3370
Week 51 = 3437
Week 52 = 3504


In [ ]:
raws1 = 20
raws2 = 19
for x in range(0,10):
  totals = (raws1*x)+(raws2*x)
  print( x + totals)

0
40
80
120
160
200
240
280
320
360


While loops ->

In [ ]:
step = 0
while step < 10:   # Command for while loop  
                    # This programme will run until step reacher 9 (< 10)
  step = step + 1   # Pythin will execute the rest of the command  until step reaches 9
  print(step)

1
2
3
4
5
6
7
8
9
10


In [ ]:
step = 0
while step < 10:
  step = step +1
  print(step)       # break is the keyword to jump out of the loop
  if step == 8:      # In this case the loop will end once step=8 
    break           # The code will stop once it has done 8 loops

1
2
3
4
5
6
7
8


In [ ]:
x = 45 
y = 80         # While loop is usually used together with other variables
while x < 50 and y < 100:
  x = x + 1 
  y = y + 1 
  print(x,y)

46 81
47 82
48 83
49 84
50 85


***1: The Hello Loop***

What do you think the following code will do? First, guess what
will happen, and then run the code in Python to see if you were
right.

In [ ]:
for x in range(0, 20):
 print('hello %s' % x)
 if x < 9:
  break

hello 0


The code will say hello 0

it will run once as when x is less than 9 which is when you run x = 0



***2: Even Numbers***

Create a loop that prints even numbers until it reaches your year
of age or, if your age is an odd number, prints out odd numbers
until it reaches your age. For example, it might print out something like this:

2

4

6

8

10

12

14





In [ ]:
x = 0
while x < 100:
  x = x + 2
  print(x)
  if x == 14:
    break

2
4
6
8
10
12
14


***3: My Five Favorite Ingredients***
Create a list containing five different sandwich ingredients, such
as the following:

*ingredients = ['snails', 'leeches', 'gorilla belly-button lint',
 'caterpillar eyebrows', 'centipede toes']*

**Now create a loop that prints out the list (including the numbers):**


1 snails

2 leeches

3 gorilla belly-button lint

4 caterpillar eyebrows

5 centipede toes

In [ ]:
ingredients = ['snails', 'leeches', 'gorilla belly-button lint', 'caterpillar eyebrows', 'centipede toes']

for x in range(0,5):
  y = x 
  x = x + 1
  print( x, ingredients[y])

1 snails
2 leeches
3 gorilla belly-button lint
4 caterpillar eyebrows
5 centipede toes


***4: Your Weight on the Moon***

If you were standing on the moon right now, your weight would be
16.5 percent of what it is on Earth. You can calculate that by multiplying your Earth weight by 0.165.
If you gained a kilo in weight every year for the next 15 years,
what would your weight be when you visited the moon each year
and at the end of the 15 years? Write a program using a for loop
that prints your moon weight for each year.

In [ ]:
weight = 60
for x in range(1,16):
  print( x, (weight+x*1)*0.165)

1 10.065000000000001
2 10.23
3 10.395000000000001
4 10.56
5 10.725
6 10.89
7 11.055
8 11.22
9 11.385
10 11.55
11 11.715
12 11.88
13 12.045
14 12.21
15 12.375


# Chapter 7 - Recycling Your Code with Functions and Modules

Functions are a set of code that tell Python to do something. They are one way to reuse code — you can use functions in your
programs again and again.   ***It is like a set of instructions***

In [ ]:
def testfunc(myname):   # You start a function by writing def, then the name you want to give the 
                        # function then write the parametre (special variable) inside the bracket
 print('hello %s' % myname)      # After that you give the instructions for what the function
                                 # will do.


testfunc('mehrab')       # Now you write the function name and set values to the parametres

hello mehrab


Parametres are special variables that can only be used inside a function.

In [ ]:
def name(name1,name2):
  print('hello %s %s' %(name1,name2))

name('mehrab','rahman')

hello mehrab rahman


In [ ]:
def savings(pocket_money, paper_route, spending):
 return pocket_money + paper_route - spending   # Return is used to return the function

print(savings(10,15,20))

5


In [ ]:
def vartest():         # This is a function that uses variables instead of parametres
  var1 = 10            # However this variable still cant be used outside of the function
  var2 = 20
  return var1 + var2

print(vartest())

30


In [ ]:
var3 = 30
def vartest():
  var1 = 10
  var2 = 20
  return var1 + var2 + var3

print(vartest())


60


In [ ]:
def moonweight(weight,increase,year):
  return (weight+increase*year)*0.165

print(moonweight(60,1,15))

12.375


In [ ]:
def moonweight(weight,increase,year):
  for x in range(1,year+1):
    monw = (weight+increase*x)*0.165
    print('year %s = %s' %(x,monw))

moonweight(60,1,15)

year 1 = 10.065000000000001
year 2 = 10.23
year 3 = 10.395000000000001
year 4 = 10.56
year 5 = 10.725
year 6 = 10.89
year 7 = 11.055
year 8 = 11.22
year 9 = 11.385
year 10 = 11.55
year 11 = 11.715
year 12 = 11.88
year 13 = 12.045
year 14 = 12.21
year 15 = 12.375


***Modules***

Modules are used to group functions, variables, and other things
together into larger, more powerful programs. Some modules are
built in to Python, and you
can download other modules
separately.

In [ ]:
import time           # For time
print(time.asctime())

Sun Dec  5 05:47:54 2021


In [ ]:
import sys                               # Use this in other IDLE
print(sys.stdin.readline())              # Colab cannot use this

name = input('What is your name: ')      # Colab uses input only
print('hello %s'% name)


What is your name: m
hello m


In [ ]:


def silly_age_joke():
 age1 = input('How old are you?: ')
 age = int(age1)                    # Use int to convert the string into a integer
 if age >= 10 and age <= 13:
  print('What is 13 + 49 + 84 + 155 + 97? A headache!')
 else:
  print('Huh?')

silly_age_joke()

How old are you?: 14
Huh?


Programming puzzles have been done above  (pg 90)

# Chapter 8 - How to Use Classes and Objects

Classes are a category of things

Objects are items that can be catagorised into classes

Classes are used to organise things into classes

In [ ]:
class Things:    # Things is the name of the class
  pass        # pass to let Python know that we are not going to give any more information

class Inanimate(Things):  # If a class is part of another class it is a children
  pass                    # and the other class is the parent. In this case Inanimate is the 
                          # children of Things and Things is the parent
class Animate(Things):
  pass

class Animals(Animate):
  pass 

class Mammles(Animals):
  pass

class Dog(Mammles):
  pass


stan = Dog()       # The name of the dog is stan. This is how we add an object to a class
                   # This code tells Python to create an object in the 

In [ ]:
class Things:             # Refer to pg 95
  pass

class Inanimate(Things):
  pass

class Animate(Things):
  pass

class Animals(Animate):
  def breathe(self):     # Defining a function self parameter is a way for one function 
    print('breathing')   # in the class to call another function in the
                         # class (and in the parent class)
  def move(self):
    print('moving')

  def eat_food(self):
    print('eating food')

class Mammals(Animals):
  def feed_young_with_milk(self):
    pass

class Dog(Mammals):
  def eat_dog_food(self):           # The dog has all the characteristics of animal and mammal
    pass                            # The dog can still use the function of animals and mammal

stan = Dog()
stan.move()


ralph = Dog()     # New object
ralph.move()      



moving
moving


In [ ]:

 
class Giraffes:
  def move(self):
   print('move')
  
  def find_food(self):           # You can make a function do another function using self
    self.move()                  # More example in #1 programming puzzle in chap 8
    print("I've found food!")
    self.eat_food()
  def eat_leaves_from_trees(self):
    self.eat_food()
  def dance_a_jig(self):
    self.move()
    self.move()
    self.move()
    self.move()

reginald = Giraffes()
reginald.dance_a_jig()

move
move
move
move


Initializing an object

We initialize an object when we are getting it ready to be used

e.g. when you want to set the number of spots on the dog, we need to initialize it

We need to create a __init __function

The init function is a way to set the properties for an
object when the object is first created, and Python will automatically call this function when we create a new object.

In [ ]:
class Dog:
  def __init__(self,spots):
    self.dog_spots = spots

tom = Dog(100)
print(tom.dog_spots)

dan = Dog(57)
print(dan.dog_spots)

100
57


Programming puzzles - pg 107

***1: The Giraffe Shuffle***

>Add functions to the Giraffes class to move the giraffe’s left and
right feet forward and backward. A function for moving the left
foot forward might look like this:

def left_Foot_Forward(self):

print('left foot forward')

>Then create a function called dance to teach Reginald to dance
(the function will call the four foot functions you’ve just created).
The result of calling this new function will be a simple dance:

reginald = Giraffes()

reginald.dance()

left foot forward

left foot back

right foot forward

right foot back

left foot back

right foot back

right foot forward

left foot forward

In [ ]:

class Giraffes:
  def left_foot_forward(self):
    print('left foot forward')

  def left_foot_back(self):
    print('left foot back')

  def right_foot_forward(self):
    print('right foot forward')

  def right_foot_back(self):
    print('right foor back')


  def dance(self):
    self.left_foot_forward()
    self.left_foot_back()
    self.right_foot_forward()
    self.right_foot_back()
    self.left_foot_back()
    self.right_foot_back()
    self.right_foot_forward()
    self.left_foot_forward()

reginald = Giraffes()
reginald.dance()

left foot forward
left foot back
right foot forward
right foor back
left foot back
right foor back
right foot forward
left foot forward


# Chapter 9 - Python’s Built-in Functions

Python has a large number of functions and modules that are ready-made for use. Python's built in functions do not need to be imported and they are already available inside colab

***The abs Function***

The abs function returns the absolute value of a number (value of a number without itds sign), which is
the value of a number without its sign. For example, the absolute
value of 10 is 10, and the absolute value of –10 is 10.
To use the abs function, simply call it with a number or variable as its parameter, like this:

In [1]:
print(abs(10))

10


In [2]:
print(abs(-10))

10


Example of usage -> When counting the total number to steps taken in a game

In [4]:
steps = -3          # When moving 3 steps left

if abs(steps) > 0:      # If not using abs, if steps < 0 or steps > 0:
  print('Moving')       #                     print('Moving')

Moving


In [5]:
steps = 10          # When moving 10 steps right

if abs(steps) > 0:
  print('Moving')

Moving


***The bool Function***

   The name bool is short for Boolean, the word programmers use to
describe a type of data that can have one of two possible values,
usually either true or false.

   The bool function takes a single parameter and returns either
True or False based on its value. When using bool for numbers, 0
returns False, while any other number returns True. Here’s how
you might use bool with various numbers:

In [6]:
print(bool(0))

False


In [7]:
print(bool(124))

True


In [9]:
print(bool(-12))

True


In [10]:
print(bool(15.67))

True


When you use bool for other values, like strings, it returns False
if there’s no value for the string (in other words, the keyword None
or an empty string). Otherwise, it will return True, as shown here:

In [18]:
print(bool())

False


In [11]:
print(bool(None))

False


In [13]:
print(bool('hello'))

True


In [1]:
print(bool(' '))

True


Example of usage

In [12]:
year = input('Year of birth: ')
if not bool(year.rstrip()):      # If you press enter without typing anything
 print('You need to enter a value for your year of birth')

Year of birth: 
You need to enter a value for your year of birth


***The dir Function***

The dir function (short for directory) returns information about any
value.

 Basically, it tells you the functions that can be used with
that value in alphabetical order.


For example, to display the functions that are available for a
list value, enter this:

In [13]:
dir(['a', 'short', 'list'])

['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__delitem__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__gt__',
 '__hash__',
 '__iadd__',
 '__imul__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__reversed__',
 '__rmul__',
 '__setattr__',
 '__setitem__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'append',
 'clear',
 'copy',
 'count',
 'extend',
 'index',
 'insert',
 'pop',
 'remove',
 'reverse',
 'sort']

   The dir function works on pretty much anything, including strings, numbers, functions, modules, objects, and classes.

But sometimes the information it returns may not be very useful.

For example, if you call dir on the number 1, it displays a number
of special functions (those that start and end with underscores)
used by Python itself, which isn’t really useful (you can usually
ignore most of them):

In [16]:
dir(1)

['__abs__',
 '__add__',
 '__and__',
 '__bool__',
 '__ceil__',
 '__class__',
 '__delattr__',
 '__dir__',
 '__divmod__',
 '__doc__',
 '__eq__',
 '__float__',
 '__floor__',
 '__floordiv__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getnewargs__',
 '__gt__',
 '__hash__',
 '__index__',
 '__init__',
 '__init_subclass__',
 '__int__',
 '__invert__',
 '__le__',
 '__lshift__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__neg__',
 '__new__',
 '__or__',
 '__pos__',
 '__pow__',
 '__radd__',
 '__rand__',
 '__rdivmod__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rfloordiv__',
 '__rlshift__',
 '__rmod__',
 '__rmul__',
 '__ror__',
 '__round__',
 '__rpow__',
 '__rrshift__',
 '__rshift__',
 '__rsub__',
 '__rtruediv__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__truediv__',
 '__trunc__',
 '__xor__',
 'bit_length',
 'conjugate',
 'denominator',
 'from_bytes',
 'imag',
 'numerator',
 'real',
 'to_bytes']

The dir function can be useful when you have a variable and
quickly want to find out what you can do with it. 

For example, run dir using the variable popcorn containing a string value, and you
get the list of functions provided by the string class (all strings are
members of the string class):

In [18]:
popcorn = 'i like popcorn'
dir(popcorn)

['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmod__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'capitalize',
 'casefold',
 'center',
 'count',
 'encode',
 'endswith',
 'expandtabs',
 'find',
 'format',
 'format_map',
 'index',
 'isalnum',
 'isalpha',
 'isascii',
 'isdecimal',
 'isdigit',
 'isidentifier',
 'islower',
 'isnumeric',
 'isprintable',
 'isspace',
 'istitle',
 'isupper',
 'join',
 'ljust',
 'lower',
 'lstrip',
 'maketrans',
 'partition',
 'replace',
 'rfind',
 'rindex',
 'rjust',
 'rpartition',
 'rsplit',
 'rstrip',
 'split',
 'splitlines',
 'startswith',
 'strip',
 'swapcase',
 'title',
 'translate',
 'upper',


At this point, you could use help to get a short description of
any function in the list.

 Here’s an example of running help against
the upper function:

In [21]:
help(popcorn.upper)

Help on built-in function upper:

upper() method of builtins.str instance
    Return a copy of the string converted to uppercase.



This code will return a string to uppercase

***The eval Function***

The eval function (short for evaluate) takes a string as a parameter
and runs it as though it were a Python expression. 

For example,
eval('print("wow")') will actually run the statement print("wow").

 **NOTE**: The eval function works only with simple expressions, such as
the following:

In [22]:
eval('10*5')

50

The eval function is often used to turn user input into Python
expressions.

 For example, you could write a simple calculator program that reads equations entered into Python and then calculates
(evaluates) the answers.


Since user input is read in as a string, Python needs to convert
it into numbers and operators before doing any calculations.

 The
eval function makes that conversion easy:

In [26]:
your_calculation = input('Enter a calculation: ')
eval(your_calculation)

Enter a calculation: 12*52


624

In this example, we use input to read what the user enters into
the variable your_calculation.

 On the next line, we enter the expression 12*52 (perhaps your age multiplied by the number of weeks
in a year). 

We use eval to run this calculation, and the result is
printed on the final line.

***The exec Function***

The exec function is like eval, except that you can use it to run more
complicated programs. 

The difference between the two is that eval
returns a value (something that you can save in a variable), whereas
exec does not. 

Here’s an example:

In [1]:
my_small_program = '''print('ham')
print('sandwich')'''
exec(my_small_program)

ham
sandwich


In the first two lines, we create a variable with a multiline
string containing two print statements, and then use exec to run
the string.


You could use exec to run mini programs that your Python program reads in from files—really, programs inside programs!

 This
can be quite useful when writing long, complicated applications.


For example, you could create a Dueling Robots game, where two
robots move around a screen and try to attack each other. 

Players
of the game would provide the instructions for their robot as mini
Python programs. 

The Dueling Robots game would read in these
scripts and use exec to run.

In [19]:
code1 = '''print('wow')
print('im surprised')'''

code2 = '''print(1+1)'''


exec(code1)
  

wow
im surprised


***The float Function***

The float function converts a
string or a number into a floatingpoint number, which is a number
with a decimal place (also called
a real number). 

For example, the
number 10 is an integer (also
called a whole number), but 10.0,
10.1, and 10.253 are all floatingpoint numbers (also called floats).

you can convert a string to float by calling float:

In [22]:
float('12')

12.0

In [25]:
float('10.93845')

10.93845

You might use float to convert values entered into your program into proper numbers, which is particularly useful when you
need to compare the value a person enters with other values.

 For
example, to check whether a person’s age is above a certain number, we could do this:

In [24]:
your_age = input('Enter your age: ')
age = float(your_age)
if age > 13:
  print('You are %s years too old' % (age - 13))

Enter your age: 20
You are 7.0 years too old


****The int Function****  ---- VERY USEFULL

The int function converts a string or a number into a whole number (or integer), which basically means that everything after the
decimal point is dropped.

---> # It can also convert a string to integer

 For example, here’s how to convert a
floating-point number into a plain integer:

In [26]:
int(123.456)

123

In [27]:
int('123')

123

However floating number string do not work and an error will pop up. (floating numbers are numbers with decimals)

It also makes the floating number into an integer if using floating numbers

In [4]:
int(12.345)

12

***The len Function***


The len function returns
the length of an object or,
in the case of a string, the
number of characters in the
string.

 For example, to get
the length of this is a test
string, you would do this:

In [5]:
len('this is a test string')

21

When used with a list or a tuple, len returns the number of
items in that list or tuple:

In [7]:
 creature_list = ['unicorn', 'cyclops', 'fairy', 'elf', 'dragon',
 'troll']

print(len(creature_list))

6


Used with a map, len also returns the number of items in
the map:

In [8]:
enemies_map = {'Batman' : 'Joker',
 'Superman' : 'Lex Luthor',
 'Spiderman' : 'Green Goblin'}

print(len(enemies_map))

3


The len function is particularly useful when you’re working
with loops.

 For example, we could use it to display the index positions of the elements in a list like this:

In [9]:
fruit = ['apple', 'banana', 'clementine', 'dragon fruit']
length = len(fruit)

for x in range(0, length):
  print('the fruit at index %s is %s' % (x, fruit[x]))

the fruit at index 0 is apple
the fruit at index 1 is banana
the fruit at index 2 is clementine
the fruit at index 3 is dragon fruit


Here, we store the length of the list in
the variable length, and then use that
variable in the range function to create our
loop. As we loop through each item
in the list, we print a message showing the
item’s index position and value. You could
also use the len function, if you had a list of
strings and wanted to print every second or
third item in the list.

***The max and min Functions***


The max function returns the largest item in a
list, tuple, or string. 

For example, here’s how
to use it with a list of numbers:

In [10]:
 numbers = [5, 4, 10, 30, 22]
 print(max(numbers))

30


A string with the characters separated by commas or spaces
will also work: 

In [11]:
strings = 's,t,r,i,n,g,S,T,R,I,N,G'
print(max(strings))

t


As this example shows, letters are ranked alphabetically, and
lowercase letters come after uppercase letters, so t is more than T.

But you don’t have to use lists, tuples, or strings. 

You can also
call the max function directly, and enter the items that you want to
compare into the parentheses as parameters:

In [12]:
print(max(10, 300, 450, 50, 90))

450


The min function works like max, except that it returns the
smallest item in the list, tuple, or string. 

Here’s our list of numbers example using min instead of max:

In [2]:
numbers = [5, 4, 10, 30, 22]
print(min(numbers))

4


Suppose you were playing a guessing game with a team of four
players, and each had to guess a number that was less than your
number.

 If any player guesses above your number, all players lose,
but if they all guess lower, they win. 

We could use max to quickly
find whether all of the guesses are lower, like so:

In [5]:
guess_this_number = 61
player_guesses = [12, 15, 70, 45]
if max(player_guesses) > guess_this_number:
  print('Boom! You all lose')

else:
  print('You win')

Boom! You all lose


In this example, we store the number to guess using the variable guess_this_number.

 The team members’ guesses are stored in
the list player_guesses.

  The if statement checks the maximum guess
against the number in guess_this_number, and if any player guesses
over the number, we print the message “Boom! You all lose.”

***The range Function***

The range function, as we’ve
seen before, is mainly used in
for loops, to loop through a section of code a specific number of
times. 

The first two parameters
given to range are called the start
and the stop. 

You saw range with
these two parameters in the earlier example of using the len
function to work with a loop.


The numbers that range generates begin with the number given
as the first parameter and end with the number that’s one less than
the second parameter. 

For example, the following shows what happens when we print the numbers that range creates between 0 and 5:

In [6]:
for x in range(0, 5):
  print(x)

0
1
2
3
4


The range function actually returns a special object called an
iterator that repeats an action a number of times. 

In this case, it
returns the next highest number each time it is called.

You can convert the iterator into a list (using the function list).

If you then print the returned value when calling range, you’ll see
the numbers it contains as well:

In [7]:
print(list(range(0, 5)))

[0, 1, 2, 3, 4]


You can also add a third parameter to range, called step.

 If the step value is not included, the number 1 is used as the step by
default.

 But what happens when we pass in the number 2 as the step? 
 
 Here’s the result:

In [8]:
count_by_twos = list(range(0, 30, 2))
print(count_by_twos)

[0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28]


Each number in the list increases by two from the previous
number, and the list ends with the number 28, which is 2 less
than 30.

 You can also use negative steps:

In [9]:
count_down_by_twos = list(range(40, 10, -2))
print(count_down_by_twos)

[40, 38, 36, 34, 32, 30, 28, 26, 24, 22, 20, 18, 16, 14, 12]


Steps are basically how many numbers it will take skip before reaching the next number. 

This can be useful when printing even/odd numbers

***The sum Function***

The sum function adds items in a list and returns the total. Here’s
an example:

In [2]:
my_list_of_numbers = list(range(0, 500, 50))
print(my_list_of_numbers)
print(sum(my_list_of_numbers))

[0, 50, 100, 150, 200, 250, 300, 350, 400, 450]
2250


On the first line, we create a list of numbers between 0 and
500, using range with a step of 50.

 Next, we print the list to see the
result. 

Finally, passing the variable my_list_of_numbers to the sum
function with print(sum(my_list_of_numbers)) adds all the items in
the list, giving the total of 2250.

***Working with Files***

Python files are the same as other files on your computer: documents, pictures, music, games . . . indeed, everything on your
computer is stored as files.

Let’s look at how to open and work with files in Python by
using the built-in function open.

 But first we need to create a
new file to play with.

To create a file in colab, open **File** then select **Upload notebook** to upload a colab file to a new or another file.

After that choose a place where you would like to upload.

***Programming Puzzles*** - pg 127

***1: Mystery Code***

What will be the result of running the following code? Guess, and
then run the code to see if you’re right.

In [3]:
a = abs(10) + abs(-10)
print(a)

b = abs(-10) + -10
print(b)

20
0


a will print out 20  -> 10 + 10

b will print out 0   -> 10 - 10

***2: A Hidden Message***

Try using dir and help to find out how to break a string into words,
and then create a small program to print every other word in the
following string, starting with the first word (this):

In [4]:
"this if is you not are a reading very this good then way you to have hide done a it message wrong"

'this if is you not are a reading very this good then way you to have hide done a it message wrong'

In [6]:
dir("this if is you not are a reading very this good then way you to have hide done a it message wrong")

['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmod__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'capitalize',
 'casefold',
 'center',
 'count',
 'encode',
 'endswith',
 'expandtabs',
 'find',
 'format',
 'format_map',
 'index',
 'isalnum',
 'isalpha',
 'isascii',
 'isdecimal',
 'isdigit',
 'isidentifier',
 'islower',
 'isnumeric',
 'isprintable',
 'isspace',
 'istitle',
 'isupper',
 'join',
 'ljust',
 'lower',
 'lstrip',
 'maketrans',
 'partition',
 'replace',
 'rfind',
 'rindex',
 'rjust',
 'rpartition',
 'rsplit',
 'rstrip',
 'split',
 'splitlines',
 'startswith',
 'strip',
 'swapcase',
 'title',
 'translate',
 'upper',


In [13]:
word= "this if is you not are a reading very this good then way you to have hide done a it message wrong"

help(word.strip)


Help on built-in function strip:

strip(chars=None, /) method of builtins.str instance
    Return a copy of the string with leading and trailing whitespace removed.
    
    If chars is given and not None, remove characters in chars instead.



In [6]:
string = "this if is you not are a reading very this good then way you to have hide done a it message wrong"

x = string.split()


for y in range(0,22,2):
  print(x[y])

this
is
not
a
very
good
way
to
hide
a
message
